### PR TITLE
Add optimization to quantum shannon decomposition

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
This adds the second optimization for the quantum shannon decomposition described in arXiv:quant-ph/0406176. It reduces the CX gate count by 4^(n-2) - 1. The cx counts before and after this pr are:

| num_qubits | QSD, current | QSD, w/ this pr | 
|---|---|---|
|3| 23  | 20|
|4| 115  | 100|
|5| 507  | 444|
|6| 2123  | 1868

### Details and comments
Adds a new decomposer class which decomposes a generic two qubit unitary, which may take 3 cnots to implement, into the product of a  diagonal gate and another unitary which takes up to 2 cnot gates. The diagonal gate of this decomposition is used to commute through neighboring gates in the QSD.

### CODE REVIEW

File: qiskit/quantum_info/synthesis/qsd.py
Comment (mtreinish) [inline qiskit/quantum_info/synthesis/qsd.py:96]: Is there a reason to not just use `two_qubit_decompose.two_qubit_decompose` here? It's already initialized as `TwoQubitBasisDecomposer(CXGate())` (although wrapped in a lazy loader).

File: qiskit/quantum_info/synthesis/qsd.py
Comment (mtreinish) [inline qiskit/quantum_info/synthesis/qsd.py:127]: We should add `opt_a2` and `_depth` to the docstring. I guess pylint doesn't complain because this is a private method, but it'd be better to document what these do.

**Review by mtreinish — COMMENTED** [review]
I still need to review the paper again to check the math used here. But assuming that's good just a few inline comments with some nits and other small things

File: qiskit/quantum_info/synthesis/qsd.py
Comment (jakelishman) [reply qiskit/quantum_info/synthesis/qsd.py:27]: You may also want to make it keyword-only (`def qs_decomposition(mat, opt_a1=True, *, _depth=0)`) to make it hard to pass accidentally.

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:54]: It's just documenting the effect of the option which I considered separate from whether it's enabled by default, which could change. One reason to keep it as an option is that there may be a special case where the user doesn't want to potentially have a 3 cx unitary pushed to beginning of their unitary.

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:127]: maybe? mentioned above

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:96]: maybe? mentioned above

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:1423]: I think this can be removed.

**Review by adjs — COMMENTED** [review]
Thanks for this PR @ewinston, 

I did not verify the decomposition. However, the number of cnots coincides exactly with the QSD (l = 2, optimized). Probably could be a nice idea to investigate what is the increment in the time to create the circuit after the optimizations.

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (israelferrazaraujo) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:1428]: Hi @adjs .
I think the reference is "Recognizing Small-Circuit Structure in Two-Qubit Operators and Timing Hamiltonians to Compute Controlled-Not Gates" https://arxiv.org/abs/quant-ph/0308045 .

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:1420]: I did  this as it was also done in the `TwoQubitWeylDecomposition.__new__()`. Perhaps it reduces load time in some cases.

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:1483]: I didn't want to eliminate theta and phi just yet. Perhaps in the future it may be useful to make these non-zero, although I don't have a concrete example now.

**Comment by ewinston** [comment]
@adjs Thanks for the review. Part of the reason for describing the function option is because it effects the structure of the circuit where I've made a choice about what direction to push the diagonal. It may be a choice that is less preferable depending on whether it reduces the chance for optimization of adjacent circuit elements.I haven't worked out pushing it the other way, but perhaps that could be added later.

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (mtreinish) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:1420]: Yeah, for a bunch of the scipy modules like this runtime imports are used to avoid the import cost on the root `import qiskit` call. Scipy imports are relatively heavy compared to the import time of the rest of qiskit.  Unless this particular method is particularly heavily used and quick so that adding a dict lookup on each call is too expensive, which I don't think it is in this case.

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py
Comment (mtreinish) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py:1417]: I updated this comment to fix the typo, but I think this is correct, if we're not using the attributes anywhere we shouldn't store them in the instance.

File: qiskit/quantum_info/synthesis/qsd.py
Comment (mtreinish) [reply qiskit/quantum_info/synthesis/qsd.py:27]: I think this change should be made to make `_depth` keyword only

**Review by mtreinish — COMMENTED** [review]
Overall I think this LGTM for the most part. One thing missing is a release note, then there were a couple inline comments that I think still should be addressed. Mainly:

* https://github.com/Qiskit/qiskit-terra/pull/8092/files#r904075456
* https://github.com/Qiskit/qiskit-terra/pull/8092/files#r905871415
* https://github.com/Qiskit/qiskit-terra/pull/8092/files#r905849287

I think after that this should be good to merge

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:27]: Thanks for the reminder. Added in 108c2a.

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:54]: I'm not sure I understand the requested test. It sounds like something already in the tests although perhaps not on a  a su4 unitary by itself. 

I updated the doc string to clarify that the reductions due to `opt_a1` and `opt_a2` are on by default in 83027a.

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:97]: I think at one time I had an option to use a different two qubit gate but that seems to be gone. This looks like a fine use of `two_qubit_decompose`. Added in 8ee787.

File: qiskit/quantum_info/synthesis/qsd.py
Comment (mtreinish) [inline qiskit/quantum_info/synthesis/qsd.py:71]: I guess pylint doesn't like documenting private keyword arguments. We can just remove this. I know you added it per my suggestion. Personally I thought it'd be better to explain what it was and that users shouldn't touch it, but we can either disable the pylint rule or just remove the arg entry. I'm fine either way.

**Review by mtreinish — COMMENTED** [review]
There are a couple pylint failures, one was caused by my suggestion(which I mentioned inline). Other than that I'm ready to approve this

File: qiskit/quantum_info/synthesis/qsd.py
Comment (ewinston) [reply qiskit/quantum_info/synthesis/qsd.py:71]: I can remove the arg entry and leave it as a comment

**Review by mtreinish — APPROVED** [review]
LGTM, assuming it passes lint again lets merge it

**Pull request merged**
